Simple stick breaking

katosh opened the pull request:
This is another attempt to introduce a new transformation of the n-simplex. The `stickbreaking` transformation is prominently used by the Dirichlet distribution as it maps the range of the Distribution (the n-simplex) to R^(n-1) where we can sample freely and apply, e.g., ADVI. The issue with the current implementation is that the transformation of later values in the vector depends on previous values. This introduces a dependency that can be confounding for ADVI and seems to produce numerical inaccuracies in some cases.

There was a previous attempt to merge the new transformation but it had a mistake in the determinant of the jacobian: #3638 

The current strikebreaking in master is an implementation of the transformation from Stan: https://mc-stan.org/docs/2_19/reference-manual/simplex-transform-section.html Which is just a repeated application of the logit transformation with adjusting range.

# Advantages
 - It is equivalent to the isometric log-ratio-transformation up to rotation. This transformation has some nice properties https://link.springer.com/article/10.1023/A:1023818214614
 - simpler implementation
 - no `eps` for numeric stability (as in https://github.com/pymc-devs/pymc3/blob/ba77d8502704e8aeb112782ee104fb339393cb19/pymc3/distributions/transforms.py#L475)
 - it transforms all values equally from R to the simplex (no bias)
 - less dependency among vector values introduced through transformation (better none-full-rank-ADVI)
 - it is numerically more stable close to the edges of the simplex, e.g., in this example:

## Current StickBreaking
```python
import pymc3 as pm
import pandas as pd

with pm.Model() as model:
    decomp = pm.Dirichlet('decomp', np.ones(10)*5e-3, shape=10)
    trace1 = pm.sample()
pd.DataFrame(trace1['decomp_stickbreaking__']).plot.kde(figsize=(10,4));
```
![image](https://user-images.githubusercontent.com/4857068/94055963-eaf8e000-fddd-11ea-8cc9-e2e800180aa4.png)

## New StickBreaking2
```python
import pymc3 as pm
import pandas as pd
from pymc3.distributions.transforms import StickBreaking2

with pm.Model() as model:
    decomp = pm.Dirichlet('decomp', np.ones(10)*5e-3, shape=10,
                          transform=StickBreaking2())
    trace2 = pm.sample()
pd.DataFrame(trace2['decomp_stickbreaking__']).plot.kde(figsize=(10,4));
```
![image](https://user-images.githubusercontent.com/4857068/94056128-2bf0f480-fdde-11ea-9271-6c9d50b2aa35.png)


The PR includes tests and there are no breaking changes as it only introduces a new transformation `pymc3.distributions.transforms.StickBreaking2` and leaves the original `pymc3.distributions.transforms.StickBreaking` untouched.

katosh committed:
simplified StickBreaking

katosh committed:
fix StickBreaking jacobian

katosh committed:
use stable logsumexp in StickBreaking

bsmith89 committed:
Rename n to Km1 to more easily compare patch.

bsmith89 committed:
Drop first dimension when computing determinant of the Jacobian of the transformation.

bsmith89 committed:
Drop newly unused expit import.

bsmith89 committed:
Side-by-side stickbreaking implementations for comparison.

bsmith89 committed:
Use same suffix for alternative stickbreaking transform.

bsmith89 committed:
Add separate tests for new stickbreaking implementation.

katosh committed:
correct jacobain of Stickbreaking2

katosh committed:
Merge remote-tracking branch 'pymc3/master' into simplify-stick-breaking

brandonwillard commented on pymc3/distributions/transforms.py:
```diff
@@ -499,6 +500,57 @@ def t_stick_breaking(eps: float) -> StickBreaking:
+        r = tt.concatenate([y+sy, tt.zeros(sy.shape)])
+        # stable according to: http://deeplearning.net/software/theano_versions/0.9.X/NEWS.html
+        sr = tt.log(tt.sum(tt.exp(r), 0, keepdims=True))
+        d = tt.log(Km1) + (Km1*sy) - (Km1*sr)
```
This is stable **after optimization**.  We're better off directly using the stable `Op` that replaces the unstable expression.

katosh replied:
I did not find a dedicated log-sum-exp operation in theano but I could use this: https://github.com/pymc-devs/pymc3/blob/67155fa4d000e73f99d761970ece73c2236a97d5/pymc3/math.py#L171

brandonwillard replied:
[This](https://github.com/pymc-devs/Theano-PyMC/blob/master/theano/tensor/opt.py#L6772) is the optimization that Theano performs, which, yes, is pretty similar to `logsumexp`.

katosh replied:
@brandonwillard Thank you for the hint! I just used the existing pymc3 implementation so it can be changed centrally.

katosh committed:
use `pymc3.math.logsumexp` in `StickBreakin2`

katosh committed:
remove old deprecated comment in `StickBreaking2`

katosh commented:
I will remove the NumPy implementation of the backward transformation `StickBreaking2.backards_val` since it is not tested and not implemented for any other transformation.

katosh committed:
remove `distributions.transforms.StickBreaking2.backwards_val`

katosh commented:
I investigated sampling divergencies in the examples above. I changed the parameter for the Dirichlet distribution to `np.ones(10)*1e-2` so not all NUTS samples diverge. Then I used the `pairplot_divergence` analogous to https://docs.pymc.io/notebooks/Diagnosing_biased_Inference_with_Divergences.html.
```python
import matplotlib.pyplot as plt

def pairplot_divergence(trace, var1, var2, i1=0, i2=0):
    v1 = trace.get_values(varname=var1, combine=True)[:, i1]
    v2 = trace.get_values(varname=var2, combine=True)[:, i2]
    _, ax = plt.subplots(1, 1, figsize=(10, 5))
    ax.plot(v1, v2, 'o', color='b', alpha=.5)
    divergent = trace['diverging']
    ax.plot(v1[divergent], v2[divergent], 'o', color='r')
    ax.set_xlabel('{}[{}]'.format(var1, i1))
    ax.set_ylabel('{}[{}]'.format(var2, i2))
    ax.set_title('scatter plot between {}[{}] and {}[{}]'.format(var1, i1, var2, i2));
    return ax
```
## Current StickBreaking
```python
pairplot_divergence(trace1, 'decomp', 'decomp', i1=2, i2=3)
```
![image](https://user-images.githubusercontent.com/4857068/94149249-eda81380-fe77-11ea-9d18-805a018353c8.png)

## New StickBreaking2
```python
pairplot_divergence(trace2, 'decomp', 'decomp', i1=2, i2=3)
```
![image](https://user-images.githubusercontent.com/4857068/94149273-f6004e80-fe77-11ea-94b9-ea793f3fa7e8.png)

## Conclusion

The parameterization from `StickBreaking2` can cure divergencies and hence avoid biases in some cases.

katosh commented:
It seems `forward_val` is sometimes called with an argument `point`, e.g., here:
https://github.com/pymc-devs/pymc3/blob/ba77d8502704e8aeb112782ee104fb339393cb19/pymc3/util.py#L183
So I will include the ignored parameter in `StickBreaking2.forward_val`.

katosh committed:
include ignored parameter `point` in `StickBreaking2.forwad_val`

katosh committed:
`StickBreaking2` in the release notes

brandonwillard commented on RELEASE-NOTES.md:
```diff
@@ -13,6 +13,7 @@
 - `sample_posterior_predictive_w` can now feed on `xarray.Dataset` - e.g. from `InferenceData.posterior`. (see [#4042](https://github.com/pymc-devs/pymc3/pull/4042))
 - Add MLDA, a new stepper for multilevel sampling. MLDA can be used when a hierarchy of approximate posteriors of varying accuracy is available, offering improved sampling efficiency especially in high-dimensional problems and/or where gradients are not available (see [#3926](https://github.com/pymc-devs/pymc3/pull/3926))
 - Change SMC metropolis kernel to independent metropolis kernel [#4115](https://github.com/pymc-devs/pymc3/pull/3926))
+- alternative transformtion of the simplex `StickBreaking2` - e.g. for the `Dirichlet` distribution. [#4129](https://github.com/pymc-devs/pymc3/pull/4129)
```
There's a typo: `transformtion` -> `transformation`.  Also, `alternative` should be capitalized so that it matches the other lines.

katosh replied:
Fixed in d84f928e49a6b0099c15a7760994e9b8b063059b and reviced in 4345c5b851990bba29fe24ab4e362ecc01ebd82c.

brandonwillard reviewed (changes requested):
Overall, this alternative stickbreaking seems fine, but why wouldn't it entirely replace the old one?  

Also, if possible, this PR should add a new test that confirms one of the advantages of this transform over the old one.  Ideally, such a test wouldn't require anything as costly as sampling.  Is there a value range that demonstrates the improved numerical stability?

katosh committed:
fix release notes typo

katosh committed:
accuracy test that only `StickBreaking2` would pass

katosh committed:
replace StickBreaking with the new alternative

katosh committed:
cite isometric logration in StickBreaking

katosh committed:
update release notes

katosh committed:
update lda-advi-aevb.ipynb with new stickbreaking

katosh commented:
I included a test `test_stickbreaking_accuracy` that only the new version would pass.

I also replaced the old stickbreaking entirely. The reason I did not do that earlier was that I was not sure if there are applications for which the old version is preferred. But I can also not find an example.

Note that, since we now replaced the old stickbreaking, this PR comes with a breaking change: `Stickbreaking.__init__` does not take an argument `eps` anymore!

Updated the notebook `lda-advi-aevb.ipynb` since it used the old stickbreaking with `eps`.

katosh committed:
remove t_stick_breaking reference

katosh committed:
remove unused import

twiecki commented:
@katosh I would then still take the `eps` kwarg and raise a deprecation error.

katosh committed:
deprecation warning for `eps` in Stickbreaking

katosh committed:
clarify jacobian implementation

katosh committed:
remove revealing file system reference

katosh committed:
Revert "remove unused import"

This reverts commit 30452dbde7fb2387855d8277bc031fd11b189e80.

katosh committed:
fix stickbreaking accuracy test for 32bit

katosh committed:
remove unused import

katosh commented:
Code coverage is reduced since:
 1. The deprecation warning is not tested.
 2. The length of the completely covered code is reduced.

brandonwillard commented:
> 1. The deprecation warning is not tested.

You can add a `noqa` to that line.

MarcoGorelli commented:
> > 1. The deprecation warning is not tested.
> 
> You can add a `noqa` to that line.

why not write a test with

```python
with pytest.warns(DeprecationWarning("<warning text>")):
    <test which sets `eps` parameter>
```
which covers it?

katosh committed:
test eps deprecation of StickBreaking

katosh committed:
fix indentation

brandonwillard commented:
> why not write a test with

You can definitely do that, but we're not really testing much of our own code in this case, so it's not a particularly relevant unit test.

katosh commented:
already done the test :)

katosh committed:
Merge remote-tracking branch 'pymc3/master' into simplify-stick-breaking

katosh commented:
I tested how close to the edge of the simplex we can go before the transformation starts to break and for the cases I tested it seems to work down to the smallest float64:
```python
>>> import numpy as np
>>> from pymc3.distributions.transforms import stick_breaking
>>> a = 5e-324
>>> vec = np.array([a, a, a, 1-(3*a)]) # a point very close to the edge of the 4-simplex
>>> stick_breaking.backward(stick_breaking.forward(vec).eval()).eval()
array([5.e-324, 5.e-324, 5.e-324, 1.e+000]) # very close to vec
```

However, the same can possibly not be said about the jacobian!

katosh commented:
I investigated the jacobian by plotting its values for points `StickBreaking.forwad(Simplex(3))` with
```python
import itertools
import numpy as np
import theano
from pymc3.distributions.transforms import stick_breaking
import matplotlib.pyplot as plt
from matplotlib import cm
from mpl_toolkits.mplot3d import Axes3D

def plot_jacobian_det(line):
    xl = list()
    yl = list()
    zl = list()
    p = theano.tensor.dvector('p')
    jd = theano.function([p], stick_breaking.jacobian_det(p))

    for x, y in itertools.product(line, repeat=2):
        xl.append(x)
        yl.append(y)
        log_jacobian_det = jd(np.array([x, y]))
        zl.append(log_jacobian_det)

    x = np.stack(xl)
    y = np.stack(yl)
    z = np.stack(zl)

    fig = plt.figure()
    ax = Axes3D(fig)
    ax.plot_trisurf(x, y, z, cmap=cm.jet)
```
I looked at the values in the center of the simplex:
```python
plot_jacobian_det(np.linspace(-1, 1, 100))
```
![image](https://user-images.githubusercontent.com/4857068/94363447-74503100-00c2-11eb-9766-1d15c9fa4687.png)

and all the way to the edge
```python
plot_jacobian_det(np.linspace(-500, 500, 100))
```
![image](https://user-images.githubusercontent.com/4857068/94363463-92b62c80-00c2-11eb-8d6b-58a5c4d927d8.png)
Note that `stick_breaking.backward(np.array([-350, 350])).eval()` is `array([9.85967654e-305, 1.00000000e+000, 9.92959040e-153])` so we go as close to the edge of the simplex as possible with float64.

There are no numerical issues apparent and the log-determinant of the jacobian seems to behave as expected down to values so close to the edge of the simplex that they cannot be distinguished by the Dirichlet distribution (I belive they are mapped to the simplex befor `Dirichlet.logp` is evaluated). But this is of course not an exhaustive investigation.

twiecki commented:
@katosh This looks great and quite thorough. Is there anything missing before merging from your end?

katosh commented:
I am done so far but of course, I can do further testing if someone has a request.

twiecki commented:
I think this is great, thanks so much for the contribution!

katosh commented:
Awesome, thank you for having me be part of this project!

helmutsimon commented:
It appears that StickBreaking.forward_val is being eliminated, with no equivalent in the new version. This would concern me, as I happen to use it in a public repository. I could work around it, but perhaps there are others using it also. Is there any depreciation warning in the meantime? I only found out about this because I wanted to add a backward_val.

ricardoV94 commented:
> It appears that StickBreaking.forward_val is being eliminated, with no equivalent in the new version. This would concern me, as I happen to use it in a public repository. I could work around it, but perhaps there are others using it also. Is there any depreciation warning in the meantime? I only found out about this because I wanted to add a backward_val.

Do you mind opening a separate issue for that? This one is pretty long and the forward_val were removed for all distributions not just StickBreaking

helmutsimon commented:
> Do you mind opening a separate issue for that? This one is pretty long and the forward_val were removed for all distributions not just StickBreaking

See discourse topic [here](https://discourse.pymc.io/t/removal-of-forward-val-from-transforms/7940)
